### ISSUE TITLE
Out of memory while trying to load lots of NetworkLinks in KML file

### ISSUE THREAD

**Issue by kevinpocovi** [mapsplugin/cordova-plugin-googlemaps#2554]
This plugin is open source project, thus you can use this plugin as free!

In order to share the issue knowledges with other people, we would like to ask you to make converstations in open issue threads as much as possible.

However some people, some companies do not allow to do that, because of their NDAs or privacy policies.

Yes, we understand thier situations.

In that case, choose **"private support"**.
If you choose "**private support**", we will make a private chat room for you, then we can discuss your issues.
We don't discover your code or your application screen captures into public spaces.

### Fee

You can choose one of following options:

- only one time, ask only one issue.
  Please donate $15 USD.
  
- 1 month private support, up to 10 issues.
  Please donate $100 USD.
  
- one year private support, no cap.
  Please donate $1,000 USD.

  
### How to apply it?

Please make an issue with this template, then donate the fee amount.
 
  https://github.com/mapsplugin/cordova-plugin-googlemaps#please-support-this-plugin-activity

You don't need to discover your information at this issue thread.
After donation, we will contact the email address which is used for paypal payment.

**Comment by wf9a5m75** [comment]
Out of memory is out of memory. Mobile device has small memory than PC, and Android allows apps to limited memory.
Large KML or lots of KML mean generates lots of overlays. That's why out of memory.

**Comment by kevinpocovi** [comment]
Hello Masashi,

Thanks for your reply. I already know that. What I need to know is what is
the best way to free out the memory while the KML overlays are not visible
by removing them from the stack, and reload them when you are back into the
same part of the map.

Thanks!

**Comment by wf9a5m75** [comment]
Unfortunately this plugin does not have such a function,. Since Google Maps SDK for Android doesn't have `setMap()` method, developer can not detach/reattach overlays after adding them.

You need to create KML files based on region, then you need to manage them by yourself using `kmloverlay.remove()` and `CAMERA_MOVE_END` event.

**Comment by kevinpocovi** [comment]
Ok, got it. Thanks.

Would you be interested in developing this feature? If so, for how much
money and in how much time?

Kevin

**Comment by wf9a5m75** [comment]
No. I have no enough time.

**Comment by wf9a5m75** [comment]
`kmloverlay.getKmlData()` returns instances of overlays. You can remove some of them.

**Comment by wf9a5m75** [comment]
Or try to add `android:largeHeap="true"` into AndroidManifest.xml file.

https://stackoverflow.com/a/11275719

**Comment by kevinpocovi** [comment]
Is there any function to check which KML should be loaded based on the
visible region? If not, our approach will be loading the index that has all
the NetworkLinks and their bounds and iterate them in order to find the
regions that overlap with the actual visual region. We will attach this
function to the CAMERA_MOVE_END event. We will check the overlap by using
this concept (http://jsfiddle.net/Lqh3mjr5/). If we already loaded any KML
overlay and it's not visible anymore on the screen we will remove it and
wait until its callback is called to show the new visible region's KML
overlay.

Do you think this could work and have a good performance?

or try to add android:largeHeap="true" into AndroidManifest.xml file.
- I understand that it should help, but we have GBs of data to load and
this would be just a patch that wont work while trying to load all the data.

kmloverlay.getKmlData() returns instances of overlays. You can remove some
of them.
- This will not be necessary as I have already divided the KML's based on
region.

Thank you.

**Comment by wf9a5m75** [comment]
In order to get visible region (nearLeft, nearRight, farLeft, farRight), you need to use `map.getVisibleRegion()`.

Converting lat/lng to pixels using `map.fromLatLngToPoint()`.

Then check which overlays are displayed (or not displayed) every time when `CAMERA_MOVE_END` event is occurred.

**Comment by kevinpocovi** [comment]
"When KML file contains
<networklink><region><bounds>...</bounds></region></networklink>, this
plugin loads when the bounds is displayed....."

That is exactly what I was trying to do before but I had the problem that
it wont let me remove an overlay even when its invisible. Do you think
there is a way to modify the Plug In so it removes them? Maybe changing
some lines of code in KmlLoader or KmlOverlay?

**Comment by wf9a5m75** [comment]
Fork this repository, and modify the code by yourself.

**Comment by kevinpocovi** [comment]
I've spent the whole weekend trying to eliminate the overlay by modifying
the repository. Could you give me an idea on which part I should modify?

I've tried changing this line (
https://github.com/mapsplugin/cordova-plugin-googlemaps/blob/v2.4.6/www/KmlLoader.js#L934)
with overlay.remove();  I've also tried adding networkOverlay.destroy() but
it didnt work.

Also tried with networkOverlay.set("_loaded", false);
networkOverlay.set("_visible", false);

If you can point us in the right direction we'll be happy to do it and send
a pull request for this feature.

Thanks

P.S why did you send me a refund?

**Comment by wf9a5m75** [comment]
I think the following code is enough.

```js
overlay.remove();
networkOverlay.set("overlay", null);
```

**Comment by kevinpocovi** [comment]
Hello Masashi,

I was able to make it work with your suggested fix and some other
modifications. What im having trouble with now is after removing the
overlays they are removed from the native side but not from the Javascript
side. Currently im removing them by using $.each(map.OVERLAYS,
function(k,v){if(v._isRemoved) delete app.map.OVERLAYS[k]}) .

Also, sometimes there are some KML's overlays that should have been removed
are still visible until I zoom in in order to satisfy the LOD condition and
zoom out again to retry the removal, or by moving to another region.

Now the App wont crash by out of memory but when I remove overlays the
memory usage isnt decreasing much. After a few seconds the memory usage
rises again to the same level or even more.
